File: sdks/typescript/src/sse.ts
/**
 * Parse an SSE stream from the server.
 *
 * The server emits three event types:
 *   - `build_log`: streamed during image build, contains `{ step, type, message }`
 *   - `error`: build failed, contains `{ error: string }`
 *   - `result`: final result JSON (sandbox data, snapshot info, etc.)
 *
 * @param resp - Fetch Response with content-type text/event-stream
 * @param onLog - Callback for build log messages
 * @returns The parsed result from the "result" event
 */
export async function parseSSEStream<T>(resp: Response, onLog: (msg: string) => void): Promise<T> {
  const reader = resp.body?.getReader();
  if (!reader) throw new Error("No response body for SSE stream");

  const decoder = new TextDecoder();
  let buffer = "";
  let result: T | null = null;

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;

    buffer += decoder.decode(value, { stream: true });
    const parts = buffer.split("\n\n");
    buffer = parts.pop() ?? "";

    for (const part of parts) {
      if (!part.trim()) continue;

      let eventType = "";
      let data = "";
      for (const line of part.split("\n")) {
        if (line.startsWith("event: ")) eventType = line.slice(7);
        else if (line.startsWith("data: ")) data = line.slice(6);
      }

      if (!data) continue;

      switch (eventType) {
        case "build_log": {
          try {
            onLog(JSON.parse(data).message ?? data);
          } catch {
            onLog(data);
          }
          break;
        }
        case "error": {
          let msg = data;
          try { msg = JSON.parse(data).error ?? data; } catch {}
          throw new Error(`Build failed: ${msg}`);
        }
        case "result": {
          result = JSON.parse(data);
          break;
        }
      }
    }
  }

  if (!result) throw new Error("No result received from build stream");
  return result;
}

File: sdks/python/opencomputer/sse.py
"""Shared SSE stream parser for build log streaming."""

from __future__ import annotations

import json
from typing import Any, Callable

import httpx


async def parse_sse_stream(
    resp: httpx.Response,
    on_log: Callable[[str], None],
) -> dict[str, Any]:
    """Parse an SSE stream from the server.

    The server emits three event types:
      - build_log: streamed during image build, contains { step, type, message }
      - error: build failed, contains { error: string }
      - result: final result JSON

    Args:
        resp: httpx streaming response with content-type text/event-stream
        on_log: Callback for build log messages

    Returns:
        The parsed result dict from the "result" event

    Raises:
        RuntimeError: If the build fails or no result is received
    """
    result: dict[str, Any] | None = None
    buffer = ""

    async for chunk in resp.aiter_text():
        buffer += chunk
        events = buffer.split("\n\n")
        buffer = events[-1]

        for event in events[:-1]:
            if not event.strip():
                continue

            event_type = ""
            data = ""
            for line in event.split("\n"):
                if line.startswith("event: "):
                    event_type = line[7:]
                elif line.startswith("data: "):
                    data = line[6:]

            if not data:
                continue

            if event_type == "build_log":
                try:
                    parsed = json.loads(data)
                    on_log(parsed.get("message", data))
                except (json.JSONDecodeError, TypeError):
                    on_log(data)
            elif event_type == "error":
                msg = data
                try:
                    msg = json.loads(data).get("error", data)
                except (json.JSONDecodeError, TypeError):
                    pass
                raise RuntimeError(f"Build failed: {msg}")
            elif event_type == "result":
                result = json.loads(data)

    if not result:
        raise RuntimeError("No result received from build stream")
    return result
